Source/src/modules/ModuleEditor.h

```diff
@@ -111,5 +117,29 @@ namespace Hachiko
         WindowProject w_project;
 
         Editor::Theme::Type theme = Editor::Theme::Type::DARK;
+
+        class History
+        {
+        public:
+            Scene::Memento* Undo();
+            Scene::Memento* Redo();
+            void Save(Scene::Memento*);
+            [[nodiscard]] bool CanUndo() const;
+            [[nodiscard]] bool CanRedo() const;
+
+            void Init();
+            void CleanUp();
+
+        private:
+            size_t current_position = 0;
+            std::vector<std::shared_ptr<Scene::Memento>> mementos {};
+        };
```

Comment:
It's kinda confusing to have this inside here, and defined inside same cpp with ModuleEditor as they both have methods with same name such as `ModuleEditor::Init` and `ModuleEditor::History::Init` or `ModuleEditor::CleanUp` and `ModuleEditor::History::CleanUp`. Maybe moving History to it's separate .h and .cpp would be better as this can cause issues to other developers if they are not careful enough.

Reply:
The pattern idea is to encapsulate the history behavior, since history belong exclusively to the editor, I decided to make it a private class of that module, but I think it's ok to move it a separate .h/.cpp if you want, let me know

Reply:
to elaborate a little more, this implementation is based off nested classes, as explained in here https://refactoring.guru/design-patterns/memento#structure